### DISCUSSION TITLE
Forking mainnet for testing

### DISCUSSION THREAD

**Discussion by toliaqat** [Agoric/agoric-sdk#8916] [Show and tell]
Forking the Agoric mainnet for testing purposes allows developers to create a parallel version of the Agoric blockchain, enabling them to experiment with and evaluate changes without affecting the main network. This process can be great for testing updates, smart contracts, and new features in a controlled environment before they are deployed on the live blockchain.
We have started testing our new releases on fork instances now. The code can be found https://github.com/agoric-labs/cosmos-genesis-tinkerer

There are multiple steps involved in creating a mainfork. Here are those steps:

- Create a new mainnet follower using state-sync.
- Export the state of the follower.
- Tinker the genesis.json in this exported state to add two powerful validators.
- Import the tinkered state into a new mainnet cluster with two validators.
- Create a cosmos snapshot of this new mainnet cluster.

Currently, we have deployed mainfork on [Ollinet](https://ollinet.agoric.net/) cluster.

Note: As time progresses, the memory and disk space needed for the mainnet increase, complicating the deployment of mainfork due to its high memory demands. Presently, deploying a fork from January's snapshot requires about 50GB of memory. Looking ahead, we plan to adopt a new forking technique that will be integrated into the startup code of cosmos-sdk. This approach will use the existing database to initiate a fork, streamlining the process and eliminating the need for extensive adjustments.

**Comment by dckc** [comment]
I gather we're also tracking work in the osmosis community on state patching logic.

**Reply by toliaqat** [reply]
Yes, their code recently has landed in cosmos-sdk https://github.com/cosmos/cosmos-sdk/pull/19280
